File: 2385.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def amountOfTime(self, root: Optional[TreeNode], start: int) -> int:
        assert root is not None
        target = None
        parent = dict()
        stack = [(root, None)]
        while len(stack) != 0:
            node, p = stack.pop()
            if node.val == start:
                target = node
            parent[node] = p
            for n in (node.left, node.right):
                if n is not None:
                    stack.append((n, node))
        visited = set()
        stack = [(target, 0)]
        res = 0
        while len(stack) != 0:
            node, d = stack.pop()
            if node in visited:
                continue
            res = max(res, d)
            visited.add(node)
            for nei in (node.left, node.right, parent[node]):
                if nei is not None:
                    stack.append((nei, d + 1))
        return res
        
            

File: 2385.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
#include <unordered_map>
#include <unordered_set>

struct Solution {
    void buildParent(
            TreeNode *node,
            TreeNode *parent,
            unordered_map<TreeNode*, TreeNode*> &parentMap,
            TreeNode *&target,
            int start){
        if (node->val == start){
            target = node;
        }
        parentMap[node] = parent;
        if (node->left != nullptr){
            buildParent(node->left, node, parentMap, target, start);
        }
        if (node->right != nullptr){
            buildParent(node->right, node, parentMap, target, start);
        }
    }

    int dfs(
            TreeNode *node, 
            unordered_map<TreeNode*, TreeNode*> &parent,
            unordered_set<TreeNode*> &visited,
            int depth){
        TreeNode * arr[3] = {node->left, node->right, parent[node]};
        int res = depth;
        visited.insert(node);
        for (int i = 0; i < 3; ++i){
            if (arr[i] != nullptr && !visited.contains(arr[i])){
                res = max(res, dfs(arr[i], parent, visited, depth) + 1);
            }
        }
        return res;
    }

    int amountOfTime(TreeNode* root, int start) {
        TreeNode *target = nullptr;
        unordered_map<TreeNode*, TreeNode*> parent;
        buildParent(root, nullptr, parent, target, start);
        unordered_set<TreeNode*> visited;
        return dfs(target, parent, visited, 0);
    }
};
